src/core_plugins/vega/public/vega_editor.controller.js

```diff
@@ -0,0 +1,72 @@
+import { uiModules } from 'ui/modules';
+import hjson from 'hjson';
+import compactStringify from 'json-stringify-pretty-compact';
+
+const module = uiModules.get('kibana/vega', ['kibana']);
+module.controller('VegaEditorController', ($scope /*, kbnUiAceKeyboardModeService*/) => {
+
+  return new (class VegaEditorController {
+    constructor() {
+      $scope.aceLoaded = (editor) => {
+        editor.$blockScrolling = Infinity;
+
+        const session = editor.getSession();
+        session.setTabSize(2);
+        session.setUseSoftTabs(true);
+
+        // FIXME: enabling this service breaks ACE text wrapping on width resize
+        // kbnUiAceKeyboardModeService.initialize($scope, editor);
+
+        this.aceEditor = editor;
+      };
+
+      $scope.formatJson = (event) => {
+        this._format(event, compactStringify, {
+          maxLength: this._getCodeWidth(),
+        });
+      };
+
+      $scope.formatHJson = (event) => {
+        this._format(event, hjson.stringify, {
+          condense: this._getCodeWidth(),
+          bracesSameLine: true,
+          keepWsc: true,
+        });
+      };
+
+      // $hack: by default, Vega vis should enable auto-apply, as it is a more natural way
+      // to develop Vega visualizations, and consistent with other Vega tools and Kibana Vega plugin.
+      $scope.$parent.$parent.$parent.$parent.autoApplyEnabled = true;
+    }
+
+    _getCodeWidth() {
+      return this.aceEditor.getSession().getWrapLimit();
+    }
+
+    _format(event, stringify, opts) {
+      event.preventDefault();
+
+      // FIXME: is this the right eval method?
+      $scope.$evalAsync(() => {
+        // TODO: error handling and reporting
+        try {
+          const doc = this.aceEditor;
+          const spec = hjson.parse(doc.getValue(), { legacyRoot: false, keepWsc: true });
+          const spec2 = stringify(spec, opts);
+          doc.setValue(spec2);
+
+          // FIXME!
+          // The dirty state of the spec is not updated via ACE's setValue -> onChange -> ace-ui
+          // Repo: disable auto, use "format as ..." to change the spec, click save.
+          // Observe that stale version is saved.
+          // FIXME: Enabling this kills the editor UNDO after formatting (loss of comments with ->JSON)
```

Comment:
@spalger, do you know how this part can be fixed without doing `$scope.vis.params.spec = spec2;` ?

Reply:
I was able to accomplish what I think you want with this: https://gist.github.com/anonymous/7ee12914caabd12aabc51e1e38d43209

The trick is using `editor.getSession().doc.setValue()` rather than `session.setValue()` or `vis.params.spec` so that the edit is stored in the undo stack and treated like any other edit from the user. Unfortunately, ui-ace only respects `change` events that happen [outside of a digest loop](https://github.com/angular-ui/ui-ace/blob/master/src/ui-ace.js#L227-L235), so we have to use `$scope.$$postDigest()` to schedule the `setValue()` call until after the digest. We could also use `setTimeout(..., 0)`, but I think `$scope.$$postDigest()` communicates the intention better, even though it's not intended for external use.

Reply:
This worked great, thanks!!!